flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CoalesceFunctionITCase.java

```diff
@@ -0,0 +1,87 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.functions;
+
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+
+import org.junit.runners.Parameterized;
+
+import java.util.Arrays;
+import java.util.List;
+
+import static org.apache.flink.table.api.DataTypes.BIGINT;
+import static org.apache.flink.table.api.DataTypes.INT;
+import static org.apache.flink.table.api.DataTypes.STRING;
+import static org.apache.flink.table.api.Expressions.$;
+import static org.apache.flink.table.api.Expressions.call;
+import static org.apache.flink.table.api.Expressions.coalesce;
+
+/** Test COALESCE and its return type. * */
+public class CoalesceFunctionITCase extends BuiltInFunctionTestBase {
+
+    @Parameterized.Parameters(name = "{index}: {0}")
+    public static List<TestSpec> testData() {
+        return Arrays.asList(
+                TestSpec.forFunction(BuiltInFunctionDefinitions.COALESCE)
+                        .onFieldsWithData(null, null, 1)
+                        .andDataTypes(BIGINT().nullable(), INT().nullable(), INT().notNull())
+                        .testResult(
+                                coalesce($("f0"), $("f1")),
+                                "COALESCE(f0, f1)",
+                                null,
+                                DataTypes.BIGINT().nullable())
+                        .testResult(
+                                coalesce($("f0"), $("f2")),
+                                "COALESCE(f0, f2)",
+                                1L,
+                                DataTypes.BIGINT().notNull())
+                        .testResult(
+                                coalesce($("f1"), $("f2")),
+                                "COALESCE(f1, f2)",
+                                1,
+                                DataTypes.INT().notNull())
+                        .testResult(
+                                coalesce($("f0"), 1),
+                                "COALESCE(f0, 1)",
+                                1L,
+                                // In this case, the return type is not null because we have a
+                                // constant in the function invocation
+                                DataTypes.BIGINT().notNull())
+                        .testResult(
```

Comment:
I'm not sure we need this test either?

Reply:
This is to test the function works properly only with a single argument

Reply:
What makes one argument special? Wouldn't no arguments be the edge case to look at? In fact, should COALESCE work with  zero arguments? Currently this is allowed, right?

Reply:
It's allowed and it returns null

Reply:
- From PostgreSQL behaviour, COALESCE should have at least one arg.
- There should be validation that the args are compatible, and the way it works for PostgreSQL is left to right, so the 1st argument non-null arg is the reference one and everything else should be "castable" to that one:
```
postgres=# select COALESCE();
ERROR:  syntax error at or near ")"
LINE 1: select COALESCE();
                        ^
postgres=# select COALESCE(1,'2');
 coalesce 
----------
        1
(1 row)

postgres=# select COALESCE(1,'2', NOW());
ERROR:  COALESCE types integer and timestamp with time zone cannot be matched
LINE 1: select COALESCE(1,'2', NOW());

postgres=# select COALESCE('foo', 1, 2);
ERROR:  invalid input syntax for type integer: "foo"
LINE 1: select COALESCE('foo', 1, 2);
                        ^
```

Reply:
We could also investigate what the SQL standard says and what other implementations do; however, this isn't a new function, so we need to consider existing pipelines as well.

Reply:
> From PostgreSQL behaviour, COALESCE should have at least one arg.

@matriv About this one, I think it makes sense as restriction, and I can give you a practical reason: What should be the type if there are no args?

> There should be validation that the args are compatible, and the way it works for PostgreSQL is left to right, so the 1st non-null arg is the reference one and everything else should be "castable" to that one:

About this one, as @Airblader sad, it might break existing users of COALESCE, which now are under the "least common type" semantics. But it would be nice to check what the sql standard has to say about it

Reply:
I've just checked in the sql spec and it seems the COALESCE requires at least one argument:

```
COALESCE <left paren> <value expression> { <comma> <value expression> }... <right paren>
```